learn/json_preprocessor.go

```diff
@@ -0,0 +1,54 @@
+package learn
+
+import (
+	"io"
+	"unicode"
+)
+
+type stripControlCharactersReader struct {
+	wrapped io.Reader
+}
+
+func newStripControlCharactersReader(wrapped io.Reader) stripControlCharactersReader {
+	return stripControlCharactersReader{wrapped: wrapped}
+}
+
+// Read up to len(p) bytes, removing any unescaped control characters found.
+// Removed characters do not count toward the total bytes read.  Returns the
+// number of bytes read.
+func (r stripControlCharactersReader) Read(p []byte) (n int, err error) {
+	pIdx := 0
+
+	// Read up to len(p) bytes, then discard any control characters.  Continue
+	// reading (and discarding control characters) until p is full or there are
+	// no more bytes to read.
+	for pIdx < len(p) {
+		remaining := len(p) - pIdx
+		buf := make([]byte, remaining)
+
+		var bufN int
+		bufN, err = r.wrapped.Read(buf)
+
+		// Copy from buf to p, skipping unescaped control characters.
+		hasPrecedingSlash := false
+		for _, r := range string(buf) {
+			if unicode.IsControl(r) && !hasPrecedingSlash {
+				continue
+			}
+			if r == '\\' {
+				hasPrecedingSlash = !hasPrecedingSlash
+			}
+			pIdx += copy(p[pIdx:], string([]rune{r}))
```

Comment:
I don't think that's the case.  We only read up to the remaining free bytes in `p`.  If anything, that might cause us to read the first byte of a multi-byte character and then decode it improperly, as well as any future bytes read.  

If we get very unlucky, the first byte could decode to a control character, which we would then remove, invalidating the rest of the string.  I think that would cause the parser to fail though, which isn't any worse than what we're currently doing.

Reply:
As discussed offline, bytes in multi-byte UTF-8 code points all have their high bits set, so this isn't a concern.